File: src/main/java/net/pengtul/pengcord/mdparse/SpoilerExtension.java
package net.pengtul.pengcord.mdparse;

import com.vladsch.flexmark.html.HtmlRenderer;
import com.vladsch.flexmark.parser.Parser;
import com.vladsch.flexmark.util.data.MutableDataHolder;
import org.jetbrains.annotations.NotNull;

public class SpoilerExtension implements Parser.ParserExtension, HtmlRenderer.HtmlRendererExtension {
    private SpoilerExtension() {
    }

    public static SpoilerExtension create() {
        return new SpoilerExtension();
    }

    @Override
    public void rendererOptions(@NotNull MutableDataHolder options) {

    }

    @Override
    public void extend(HtmlRenderer.@NotNull Builder htmlRendererBuilder, @NotNull String rendererType) {
        if (htmlRendererBuilder.isRendererType("HTML")) {
            htmlRendererBuilder.nodeRendererFactory(new SpoilerNodeRenderer.Factory());
        }
    }

    @Override
    public void parserOptions(MutableDataHolder options) {

    }

    @Override
    public void extend(Parser.Builder parserBuilder) {
        parserBuilder.customDelimiterProcessor(new SpoilerDelimiterProcessor());
    }
}

File: src/main/kotlin/net/pengtul/pengcord/mdparse/SpoilerExtension.kt
package net.pengtul.pengcord.mdparse

import com.vladsch.flexmark.ext.gfm.strikethrough.internal.StrikethroughJiraRenderer
import com.vladsch.flexmark.ext.gfm.strikethrough.internal.StrikethroughNodeRenderer
import com.vladsch.flexmark.ext.gfm.strikethrough.internal.StrikethroughYouTrackRenderer
import com.vladsch.flexmark.html.HtmlRenderer
import com.vladsch.flexmark.parser.Parser
import com.vladsch.flexmark.util.data.MutableDataHolder
import com.vladsch.flexmark.util.data.NullableDataKey


class SpoilerExtension: Parser.ParserExtension, HtmlRenderer.HtmlRendererExtension {
    fun create(): SpoilerExtension {
        return SpoilerExtension()
    }

    override fun parserOptions(options: MutableDataHolder?) {
    }

    override fun extend(parserBuilder: Parser.Builder?) {
        parserBuilder?.customDelimiterProcessor(SpoilerDelimiterProcessor())
    }

    override fun rendererOptions(options: MutableDataHolder) {
    }

    override fun extend(htmlRendererBuilder: HtmlRenderer.Builder, rendererType: String) {
        if (htmlRendererBuilder.isRendererType("HTML")) {
            htmlRendererBuilder.nodeRendererFactory(SpoilerNodeRenderer.Companion.Factory())
        }
    }
}